### PR TITLE
OASIS Stable Diffusion - Alexander Nguyen 46429515

### PR DESCRIPTION
**Pull request by anguyen464**
This is a pull request for a stable diffusion model trained on the OASIS brain dataset to generate high-resolution images of brains. The model was implemented via Pytorch, using a simple U-Net model which contains a downsampling layer and an upsampling layer. Both layers consist of convolution layers, batch normalization layers, and a rectified linear unit activation function. During the training process, it was noted that the model resulted in a loss drop from approximately 0.8 to 0.1 after a few epochs (further details in the README file). 

Files added:
* `README.md` - Documentation on the model, noting its usage and results of the training and generated images
* `dataset.py` - Creates a data loader using the OASIS data set and preprocesses the data for training
* `module.py` - Stable Diffusion model using a simple U-Net model for the Denoising process
* `predict.py` - Loads the saved model parameters and generates images from random noise images
* `train.py` - Trains the Stable Diffusion model, saving the parameters every 5 epochs (if conditions are met)
* `utils.py` - Helper Python file, holds hyperparameters, pre-calculated values, and helper functions for other Python files
*  `train_predict.sh` - Bash script used to run train.py and predict.py

Folders added:
* `image_output` - Training images output used for README documentation
* `predict_image` - Input images to be used in predict.py image generation
* `util` - Contains loss over epoch plot image used in README documentation

### CODE REVIEW

**Comment by SiyuLiu0329** [comment]
**This is an initial inspection, no action is required at this point**

Difficulty: Hard

Readme:
- Project Overview: Good
- Model: Good
- Data + Preprocessing: Good
- Usage: Good
- Training / Loss Curve(s): Good
- Result Demonstration: Good
- References: Good

Functionality:
- Generation: Not quite working but fair attempt. I don think this is stable diffusion, this looks like a regular diffusion model

Code:
- Consistent with Results: Yes
- File Structure: Good
- Commenting: Good
- Commit frequency: Good
- Commit messages: Good

Other Comments:
N/A

**Comment by shakes76** [comment]
# Marking
## Good Practice (Design/Commenting, TF/Torch Usage)
Adequate design and implementation
Good spacing and comments
Header blocks missing -1
## Recognition Problem 
Solves problem 
Driver Script present
File structure present
Shows Usage & Demo & Visualisation & Data usage
Module present
Commenting
No Data leakage
Difficulty: Hard
## Commit Log
Meaningful commit messages
Progressive commits used
## Documentation
ReadMe good
Model/technical explanation
Good Description and Comments 
Markdown used and PDF submitted
## Pull Request
Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch) 
No Feedback required
Request Description good
